### ISSUE TITLE
OS X pkg installer does not inform user when it doesn't install anything due to a newer version already being installed

### ISSUE THREAD

**Issue by DustinCampbell** [dotnet/cli#1767]
## Steps to reproduce

1. Install the latest .NET CLI beta bits on OS X
2. Download and install the most recent stable drop from https://dotnet.github.io/getting-started/

## Expected  behavior

Installer should inform the user that a newer version is installed. Bonus points if the installer offers the user the option of uninstalling the newer version, but I realize that's already a [known issue](https://github.com/dotnet/cli/blob/rel/1.0.0/Documentation/known-issues.md).

## Actual behavior

The installation appears to be successful. User tears hair out in frustration when they eventually realize what happened.

**Comment by blackdwarf** [comment]
@DustinCampbell good one. We cannot uninstall, but we should be able to detect the new version and do...something.

**Comment by piotrpMSFT** [comment]
This is a point-in-time issue. Installation will be side-by-side moving forward, so the installation of a previous version will succeed.

**Comment by blackdwarf** [comment]
@piotrpMSFT we need to check if the PKG will skip installing if it detects a newer version in general. For instance, I have a newer version on disk, but I need the older version; even though we are SxS, will the PKG detect the newer version and no-op by default?

**Comment by mellinoe** [comment]
It seems that we will want to make each side-by-side version of our installers have a unique identifier so that it is never possible to "upgrade" or "downgrade" a given version of CLI. By that, I mean add the "CLI version string" to the pkg identifier itself. This is ignoring servicing events where we might actually want to  upgrade an existing version. Apart from that, the above seems like a good "default" state for OSX installers.

**Comment by blackdwarf** [comment]
@mellinoe makes sense. It is also better because it will give the users a better way to find what was installed for a given version using `pkgutil`.

**Comment by DustinCampbell** [comment]
Perhaps I'm just not understanding, but how will side-by-side installation work? In the scenario above, what version would be picked by default if I invoke "dotnet restore" with no additional parameters. If the answer is "the highest version number", it doesn't really solve the problem I reported at all.

**Comment by blackdwarf** [comment]
@DustinCampbell the problem you reported has to do with  the way the `installer(1)` does the installation of PKGs, no?

**Comment by DustinCampbell** [comment]
It does, and @piotrpMSFT was asserting that side-by-side installation would fix the issue. I'm asking if side-by-side installation would actually change the "Actual Results" above. If "dotnet" still calls the beta version and the user expected it to call the one they installed, then side-by-side installation doesn't really change the Actual Results. The steps for the user to workaround the situation are better, but they still have to figure out what happened. Does that make sense?

**Comment by blackdwarf** [comment]
@DustinCampbell ah yes. In this case you would specify the version you need in your `global.json` file. This is still being baked so more details will come soon.

**Comment by DustinCampbell** [comment]
OK, well in that case, it wouldn't necessarily solve the experience issue. The user would still need to spelunk to find the appropriate wizardry to make things work.

In my case, I had installed (and forgotten that I'd installed) a recent beta build of the CLI that broke "dotnet restore". Realizing that was what happened and actually sorting it out took several hours. Our users might not hang in there for quite as long. :smile:

**Comment by blackdwarf** [comment]
Sure there is a documentation/experience issue here to sort through as we move closer to the goal. Would like to hear about your experiences though of sorting out the problem you encountered.

**Comment by DustinCampbell** [comment]
Experience was terrible. :smile: Finding the list of known issues was a pain since it doesn't appear to actually be linked from dotnet.github.io (why is it not on the Support page?), and "pkgutil" is definitely _not_ a pleasant tool to use.

> pkgutil --list-pkgs

Oh my, that's a big list of packages. Which one am I looking for?

> pkgutil --pkg-info com.microsoft.dotnet.cli

That's a really big list of files to delete. I hope I don't miss one and make things worse.

> sudo pkgutil --forget com.microsoft.dotnet.cli

I really hope I got everything. Once I do this, I'll have no idea what I missed.

**Comment by blackdwarf** [comment]
@DustinCampbell the known issue page being linkes is the only thing we can help with, unfortunately. The rest is just the way OS X deals with PKGs and what facilities it provides.

**Comment by DustinCampbell** [comment]
I understand that's the way .pkgs working in OS X. Vendors are expected to provide their own uninstallers, but that would be a different ask. I'm looking for a way to help users connect the dots. Because the installer left me thinking that everything was working, I ended up off in the weeds.
